AsyncNats/NatsConnection.cs

```diff
@@ -157,32 +165,67 @@ private async Task ReadWriteAsync(CancellationToken disconnectToken)
 
                 try
                 {
-                    _logger?.LogTrace("Connecting to {Server}", Options.Server);
-                    await socket.ConnectAsync(Options.Server);
+                    _currentServer = _serverPool.SelectServer(_isRetry);
+                    _logger?.LogTrace("Connecting to {Server}", _currentServer);
+
+                    IPEndPoint ipEndpoint;
+                    if (IPAddress.TryParse(_currentServer.Host, out var ipAddress))
+                    {
+                        ipEndpoint = new IPEndPoint(ipAddress, _currentServer.Port);
+                    }
+                    else
+                    {
+                        _logger?.LogTrace("Resolving dns hostname {Host}", _currentServer.Host);
+
+                        try
+                        {
+                            var resolvedHost = await (Options.DnsResolver ?? Dns.GetHostAddressesAsync)(_currentServer.Host);
+
+                            var selectedIpAddress = resolvedHost.OrderBy(i => Guid.NewGuid()).First();
```

Comment:
Shouldn't the IP be part of the server selection? [e.g. NatsServerPool]

If you have 1 hostname (e.g. nats.something.com ) with 4 servers, you'd want it to try all of these and this could potentially get stuck on the same [non-working] IP.

Reply:
@Da-Teach. My reasoning was that hostname translation to ip is a cross-cutting concern, and different system may have different semantics to what having multiple ips returned for a single hostname means. It may mean that all those ips point actually to the same machine in fact, not necessarily different endpoints from a physical perspective. I didn't want to introduce too many assumptions here, and I added the possibility to configure a custom DNS resolver that could eventually encapsulate those semantics. (round robinng on ips, or preferring same subnet ips first, or whatever requirements people may have). In fact, one of my problems with the official NATS client, is that on my system I want to favor local nats instances before falling back to remote ones, behind a consul dns resolver, so I need to create a new client instead of relying on the reconnect logic because there's no way to express that in their client.
 As for your example of getting stuck to the same ip, I believe in that case it could take a while, but eventually the random selection should fall on a working ip.

Reply:
I can see your reasoning, but I feel like in it's current implementation I do not have a lot of control over server/ip selection.

Lets say you have 3 servers, all of which have multiple IP addresses [perhaps different regions], because IP selection isn't part of ServerPool you don't really have a lot of control over which server is selected. So even if you have 3 region based clusters, you can't try all servers in a specific region before moving to a different region.

I'll think about it a bit, I feel like a "INatsServerSelector" or something would give more control. It would need a "SelectServerAsync" function which returns an IPEndPoint. The basic functionality would still be based on host [and use the above as a IP selection], but a developer could write their own INatsServerSelector implementation for more fine grained control? 

Because DNS resolve is done in a different location, even though you can change DNS resolve, you can't 'really' control host selection.

Reply:
Just as a reference, the official nats server will do a simple dns query at the string provided. I imagine that use cases requiring such fine grained control were common this option would be present there. 
They don’t even offer a custom resolver option, they simply pass the dns entry directly to tcp client.

Reply:
Yes but you added the custom resolver to at least have a bit more control over the server selection, so there's some use-case. So one can argue "why not go the extra mile" (which isn't a huge change) and allow for further control in a way that's a bit more transparent to the developer.   

Either use the basic implementation which features "round-robin based on hostname" or "completely random".  And if that doesn't satisfy the developer, they can do it however they want by implementing their own logic.

Which would be "custom server selector".   The interface would have 2 or 3 functions:
* SetConfiguredServers [or a property]
* SetDiscoveredServers [or a property]
* ValueTask<IPEndPoint> GetServerAsync(bool isReconnect)